src/renderer.ts

```diff
@@ -89,8 +96,13 @@ class BokehJSExec extends Widget implements IRenderMime.IRenderer {
       this._script_element.textContent = data
       if ((window as any).Bokeh.embed.kernels !== undefined) {
         this._document_id = metadata.id as string;
-        const kernel = this._manager.context.session.kernel;
-        (window as any).Bokeh.embed.kernels[this._document_id] = kernel;
+        const kernel = this._manager.context.session.kernel as any;
```

Comment:
Haven't found a nice way to make this not use ``any`` yet.

Reply:
That's because `_comms` and `_targetRegistry` are private attributes on `kernel`. Do we really need to expose them in our KernelProxy object for BokehJS to work?

Reply:
At least using my current approach it seems so, would be glad to be proved wrong though.

Reply:
Additionally, is `_targetRegistry` used by Holoviews? Because it's not required for vanilla Bokeh AFAICT

Reply:
``registerCommTarget`` uses it internally and since I'm just patching the method onto an object ``this`` gets overridden. It's really somewhat hacky, maybe an approach using a closure would be better.

Reply:
how about?

```
diff --git a/src/renderer.ts b/src/renderer.ts
index 0c36cd5..145b943 100644
--- a/src/renderer.ts
+++ b/src/renderer.ts
@@ -15,6 +15,10 @@ import {
   ReadonlyJSONObject
 } from '@phosphor/coreutils'

+import {
+  IDisposable
+} from '@phosphor/disposable';
+
 import {
   Widget
 } from '@phosphor/widgets'
@@ -25,12 +29,12 @@ import {


 export declare interface KernelProxy {
-  _comms: {[key: string]: Promise<Kernel.IComm>},
-  _targetRegistry: {[key: string]: (comm: Kernel.IComm) => void},
-  registerCommTarget: (target: string, fn: (comm: Kernel.IComm) => void) => void,
+  // copied from https://github.com/jupyterlab/jupyterlab/blob/master/packages/services/src/kernel/default.ts#L605
+  registerCommTarget(targetName: string, callback: (comm: Kernel.IComm, msg: KernelMessage.ICommOpenMsg) => void): IDisposable,
 }


+
 /**
  * The MIME types for BokehJS.
  */
@@ -96,11 +100,11 @@ class BokehJSExec extends Widget implements IRenderMime.IRenderer {
       this._script_element.textContent = data
       if ((window as any).Bokeh.embed.kernels !== undefined) {
         this._document_id = metadata.id as string;
-        const kernel = this._manager.context.session.kernel as any;
+        const registerClosure = (targetName: string, callback: (comm: Kernel.IComm, msg: KernelMessage.ICommOpenMsg) => void): IDisposable => {
+          return this._manager.context.session.kernel.registerCommTarget(targetName, callback)
+        }
         const kernel_proxy: KernelProxy = {
-          _comms: kernel._comms,
-          _targetRegistry: kernel._targetRegistry,
-          registerCommTarget: kernel.registerCommTarget
+          registerCommTarget: registerClosure
         };
         (window as any).Bokeh.embed.kernels[this._document_id] = kernel_proxy;
         this._manager.context.session.statusChanged.connect((session: IClientSession, status: Kernel.Status) => {
```

Reply:
Looks great, I had something similar but didn't think to look up the callback signature.

Reply:
Just tested it and it seems to work.